presto-druid/src/main/java/com/facebook/presto/druid/DruidBrokerPageSource.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.druid;
+
+import com.facebook.airlift.json.ObjectMapperProvider;
+import com.facebook.presto.druid.DruidQueryGenerator.GeneratedDql;
+import com.facebook.presto.spi.ColumnHandle;
+import com.facebook.presto.spi.ConnectorPageSource;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.type.BigintType;
+import com.facebook.presto.spi.type.DoubleType;
+import com.facebook.presto.spi.type.RealType;
+import com.facebook.presto.spi.type.TimestampType;
+import com.facebook.presto.spi.type.Type;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import org.joda.time.DateTime;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.ISODateTimeFormat;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.druid.DruidErrorCode.DRUID_BROKER_RESULT_ERROR;
+import static com.facebook.presto.util.DateTimeZoneIndex.getChronology;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class DruidBrokerPageSource
+        implements ConnectorPageSource
+{
+    private final ObjectMapper objecMapper = new ObjectMapperProvider().get();
+    private final ConnectorSession session;
+    private final GeneratedDql brokerDql;
+    private final List<ColumnHandle> columnHandles;
+    private final DruidClient druidClient;
+
+    private boolean finished;
+    private long readTimeNanos;
+    private long completedBytes;
+    private long completedPositions;
+
+    public DruidBrokerPageSource(
```

Comment:
This patch has many pinot-related concept. Is broker something for druid as well?

Reply:
oh, yes, Druid has broker, too: https://druid.apache.org/docs/latest/design/broker.html